File: hw/medina/u1/hw14SafeInJsonChickenFarmSimulator/javaScript/view/FarmSimulator_view.js
const fs = require("fs");
const readline = require("readline");

const FARM_FILE_PATH = "./farm.json";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function ask(question) {
  return new Promise(resolve => rl.question(question, resolve));
}

let coops = [];
let farmerName = "Farmer";

function loadData() {
  if (fs.existsSync(FARM_FILE_PATH)) {
    const data = fs.readFileSync(FARM_FILE_PATH, "utf-8");
    coops = JSON.parse(data);
    console.log("Loaded existing farm data.");
  } else {
    console.log("No data found. Creating new farm file...");
    coops = [];
    saveData();
  }
}

function saveData() {
  fs.writeFileSync(FARM_FILE_PATH, JSON.stringify(coops, null, 2));
}

function printMenu() {
  console.log(`\n--- ${farmerName}'s Chicken Farm Simulator ---`);
  console.log("1. List all coops and chickens");
  console.log("2. Add a new chicken");
  console.log("3. Add a new coop");
  console.log("4. Run farm simulation");
  console.log("5. Edit a chicken");
  console.log("6. Delete a chicken");
  console.log("7. Exit");
}

async function listCoopsAndChickens() {
  console.log("\n--- Farm Status ---");
  if (coops.length === 0) {
    console.log("There are no coops on this farm.");
    return;
  }

  for (const coop of coops) {
    console.log(`\n ${coop.location} (ID: ${coop.id}) contains:`);
    if (!coop.chickens || coop.chickens.length === 0) {
      console.log(" - (Empty)");
    } else {
      for (const chicken of coop.chickens) {
        console.log(` - ${chicken.name} (ID: ${chicken.id})`);
      }
    }
    console.log("------------------------------------");
  }
}

async function addCoop() {
  console.log("\n--- Add New Coop ---");
  const id = parseInt(await ask("Enter coop ID: "));
  if (coops.some(c => c.id === id)) {
    console.log("Error: Coop with this ID already exists.");
    return;
  }
  const location = await ask("Enter coop location: ");

  const newCoop = { id, location, chickens: [] };
  coops.push(newCoop);
  saveData();
  console.log(`Successfully added coop at ${location}!`);
}

async function addChicken() {
  console.log("\n--- Add New Chicken ---");
  if (coops.length === 0) {
    console.log("You must add a coop first!");
    return;
  }

  const id = parseInt(await ask("Enter chicken ID: "));
  if (coops.some(c => c.chickens.some(ch => ch.id === id))) {
    console.log("Error: A chicken with this ID already exists.");
    return;
  }

  const name = await ask("Enter chicken name: ");
  const color = await ask("Enter chicken color: ");
  const age = parseInt(await ask("Enter chicken age: "));
  const isMoltingInput = await ask("Is the chicken molting? (true/false): ");
  const isMolting = isMoltingInput.toLowerCase() === "true";

  console.log("Available coops:");
  coops.forEach(c => console.log(` - ID ${c.id}: ${c.location}`));
  const coopId = parseInt(await ask("Enter the ID of the coop to add this chicken to: "));
  const coop = coops.find(c => c.id === coopId);

  if (!coop) {
    console.log("Error: Coop not found.");
    return;
  }

  const newChicken = { id, name, color, age, isMolting };
  coop.chickens.push(newChicken);
  saveData();
  console.log(`Successfully added ${name} to coop ${coop.location}!`);
}

function runSimulation() {
  console.log("\n--- Running Farm Simulation ---");
  for (const coop of coops) {
    for (const chicken of coop.chickens) {
      console.log(`Chicken ${chicken.name} (ID ${chicken.id}) in ${coop.location} is clucking, eating, and wandering.`);
    }
  }
  console.log("Simulation finished successfully.");
}

async function editChicken() {
  console.log("\n--- Edit Chicken ---");
  if (coops.length === 0) {
    console.log("No coops available.");
    return;
  }

  await listCoopsAndChickens();
  const chickenId = parseInt(await ask("Enter the ID of the chicken to edit: "));

  const coop = coops.find(c => c.chickens.some(ch => ch.id === chickenId));
  if (!coop) {
    console.log("Chicken not found.");
    return;
  }

  const chicken = coop.chickens.find(ch => ch.id === chickenId);

  const newName = await ask(`Enter new name (${chicken.name}): `) || chicken.name;
  const newColor = await ask(`Enter new color (${chicken.color}): `) || chicken.color;
  const newAgeInput = await ask(`Enter new age (${chicken.age}): `);
  const newAge = newAgeInput ? parseInt(newAgeInput) : chicken.age;
  const newMoltingInput = await ask(`Is molting? (${chicken.isMolting}) (true/false): `);
  const newMolting = newMoltingInput ? newMoltingInput.toLowerCase() === "true" : chicken.isMolting;

  chicken.name = newName;
  chicken.color = newColor;
  chicken.age = newAge;
  chicken.isMolting = newMolting;

  saveData();
  console.log("Chicken updated successfully.");
}

async function deleteChicken() {
  console.log("\n--- Delete Chicken ---");
  if (coops.length === 0) {
    console.log("No coops available.");
    return;
  }

  await listCoopsAndChickens();
  const chickenId = parseInt(await ask("Enter the ID of the chicken to delete: "));

  const coop = coops.find(c => c.chickens.some(ch => ch.id === chickenId));
  if (!coop) {
    console.log("Chicken not found.");
    return;
  }

  coop.chickens = coop.chickens.filter(ch => ch.id !== chickenId);
  saveData();
  console.log("Chicken deleted successfully.");
}

async function main() {
  console.log("Welcome to the Chicken Farm Simulator");
  farmerName = await ask("Please enter the farmer's name: ");
  loadData();

  let running = true;
  while (running) {
    printMenu();
    const choice = await ask("Enter your choice (1-7): ");

    switch (choice) {
      case "1":
        await listCoopsAndChickens();
        break;
      case "2":
        await addChicken();
        break;
      case "3":
        await addCoop();
        break;
      case "4":
        runSimulation();
        break;
      case "5":
        await editChicken();
        break;
      case "6":
        await deleteChicken();
        break;
      case "7":
        running = false;
        console.log("Saving data and exiting. Goodbye!");
        break;
      default:
        console.log("Invalid choice. Please enter a number between 1 and 7.");
    }
  }

  rl.close();
}

main();

File: hw/medina/u1/hw14SafeInJsonChickenFarmSimulator/python/view/farm_simulator_view.py
from model.farm import Farm
from model.coop import Coop
from model.chicken import Chicken

class FarmSimulatorView:
    def __init__(self, farmer_name):
        self.farmer_name = farmer_name
        self.farm = Farm("data/farm.json")

    def print_menu(self):
        print(f"\n--- {self.farmer_name}'s Chicken Farm Simulator ---")
        print("1. List all coops and chickens")
        print("2. Add a new chicken")
        print("3. Add a new coop")
        print("4. Run farm simulation")
        print("5. Edit a chicken")
        print("6. Delete a chicken")
        print("7. Exit")

    def list_coops_and_chickens(self):
        print("\n--- Farm Status ---")
        coops = self.farm.list_coops()
        if not coops:
            print("There are no coops on this farm.")
            return
        for coop in coops:
            print(coop)

    def add_coop(self):
        coop_id = int(input("Enter coop ID: "))
        location = input("Enter coop location: ")
        self.farm.add_coop(Coop(coop_id, location))

    def add_chicken(self):
        if not self.farm.coops:
            print("You must add a coop first!")
            return
        chicken_id = int(input("Enter chicken ID: "))
        name = input("Enter chicken name: ")
        color = input("Enter chicken color: ")
        age = int(input("Enter chicken age: "))
        is_molting = input("Is molting? (true/false): ").lower() == "true"
        print("Available coops:")
        for c in self.farm.coops:
            print(f" - ID {c.id}: {c.location}")
        coop_id = int(input("Enter coop ID: "))
        chicken = Chicken(chicken_id, name, color, age, is_molting)
        self.farm.add_chicken(coop_id, chicken)

    def run_simulation(self):
        print("\n--- Running Farm Simulation ---")
        for coop in self.farm.coops:
            for chicken in coop.chickens:
                print(f"Chicken {chicken.name} (ID {chicken.id}) in {coop.location} is clucking, eating, and wandering.")
        print("Simulation finished successfully.")

    def edit_chicken(self):
        chicken_id = int(input("Enter chicken ID to edit: "))
        name = input("New name (leave blank to keep current): ")
        color = input("New color (leave blank to keep current): ")
        age_input = input("New age (leave blank to keep current): ")
        is_molting_input = input("Is molting? (true/false, leave blank to keep current): ")

        kwargs = {}
        if name: kwargs["name"] = name
        if color: kwargs["color"] = color
        if age_input: kwargs["age"] = int(age_input)
        if is_molting_input: kwargs["is_molting"] = is_molting_input == "true"

        self.farm.edit_chicken(chicken_id, **kwargs)

    def delete_chicken(self):
        chicken_id = int(input("Enter chicken ID to delete: "))
        self.farm.delete_chicken(chicken_id)

    def run(self):
        running = True
        while running:
            self.print_menu()
            choice = input("Enter your choice (1-7): ")
            if choice == "1":
                self.list_coops_and_chickens()
            elif choice == "2":
                self.add_chicken()
            elif choice == "3":
                self.add_coop()
            elif choice == "4":
                self.run_simulation()
            elif choice == "5":
                self.edit_chicken()
            elif choice == "6":
                self.delete_chicken()
            elif choice == "7":
                print("Saving data and exiting. Goodbye!")
                running = False
            else:
                print("Invalid choice. Please try again.")
